Lead Scoring Demo for March 2025 Webinar (#286)
confluentinc/demo-scene: confluent-lead-scoring-sample/README.md

Please refer to this article for step by step details: https://medium.com/@angusll/deploying-an-deep-learning-model-using-kserve-and-google-cloud-vertex-ai-ae2850caf27
4. Deploy model to Vertex AI Endpoint
After the model is imported, you can deploy it to an endpoint for online serving. Similarly, please refer to the article above for step by step details.
5. Once you've done deploying the model to an endpoint, you can send it a POST request to verify if it's working as expected.
```
curl \
-X POST \
-H "Authorization: Bearer $(gcloud auth print-access-token)" \
-H "Content-Type: application/json" \
"https://us-central1-aiplatform.googleapis.com/v1/projects/${PROJECT_ID}/locations/us-central1/endpoints/${ENDPOINT_ID}:predict" \
-d "@${INPUT_DATA_FILE}"
```
## KServe
In this project, KServe is used to build the model server for the following reasons:

Flexibility. If the default Serving Runtime doesn't meet your requirements, you can create your own model server using the KServe ModelServer API to deploy it as a Custom Serving Runtime on KServe. It offers great flexibility that caters to the various goals we hope to achieve with the model server.

> Simplicity. KServe offers machine learning-specific template methods like preprocess, predict, and postprocess, allowing you to seamlessly deploy your model once you inherit them correctly.

Reviewer: ```suggestion
Simplicity. KServe offers machine learning-specific template methods like preprocess, predict, and postprocess, allowing you to seamlessly deploy your model once you inherit them correctly.

```

All you need to do is:

Subclass `kserve.Model`.

Override the `load`, `preprocess`, and `predict` methods.

Below is a code sample of how to implement your custom model based on the two steps above. This is the only Python file that you need in this project.

```
from kserve import Model, ModelServer
from typing import Dict

class YourModel(Model):
    def __init__(self, name: str):
        super().__init__(name)
        self.name = name
        self.load()

    def load(self):
        # implement your load method

    def preprocess(self):
        # implement your preprocess method

    def predict(self, request: Dict) -> Dict:
        # implement your predict method

if __name__ == "__main__":
    model = YourModel("custom-model")
    kserve.ModelServer().start([model])
```

> Here're some useful links: https://kserve.github.io/website/latest/modelserving/v1beta1/custom/custom_model/

Author: ```suggestion
## Additional Resources

- [Companion blog post](https://www.confluent.io/blog/supercharge-lead-scoring-with-apache-flink-and-google-cloud-vertex-ai/) to this demo
- [How to deploy a custom Python serving runtime in KServe](https://kserve.github.io/website/latest/modelserving/v1beta1/custom/custom_model/)
- [KServe Model Source Code](https://github.com/kserve/kserve/blob/master/python/kserve/kserve/model.py#L183)```

> https://github.com/kserve/kserve/blob/37be2411b462687026686abac8d7bdb7c79c9cd2/python/kserve/kserve/model.py#L157

Author: ```suggestion
```

> For more details, feel free to check out my blog: https://www.confluent.io/blog/supercharge-lead-scoring-with-apache-flink-and-google-cloud-vertex-ai/

Reviewer: Rewrite this to be clearer and not require that people click to see what is at the links, e.g.:

```
## Additional Resources

- [Companion blog post](https://www.confluent.io/blog/supercharge-lead-scoring-with-apache-flink-and-google-cloud-vertex-ai/) to this demo
- [How to deploy a custom Python serving runtime in KServe](https://kserve.github.io/website/latest/modelserving/v1beta1/custom/custom_model/)
- [fill in what this is](https://github.com/kserve/kserve/blob/37be2411b462687026686abac8d7bdb7c79c9cd2/python/kserve/kserve/model.py#L157)

Reviewer 2: Thank you for pointing it out, will do

> For more details, feel free to check out my blog: https://www.confluent.io/blog/supercharge-lead-scoring-with-apache-flink-and-google-cloud-vertex-ai/

Author: ## Additional Resources

- [Companion blog post](https://www.confluent.io/blog/supercharge-lead-scoring-with-apache-flink-and-google-cloud-vertex-ai/) to this demo
- [How to deploy a custom Python serving runtime in KServe](https://kserve.github.io/website/latest/modelserving/v1beta1/custom/custom_model/)
- [KServe Model Source Code](https://github.com/kserve/kserve/blob/master/python/kserve/kserve/model.py#L183)

> For more details, feel free to check out my blog: https://www.confluent.io/blog/supercharge-lead-scoring-with-apache-flink-and-google-cloud-vertex-ai/

Author: ```suggestion
```
